main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -0,0 +1,170 @@
+          case Expression.False => exp2
+
+          // Don't know one way or the other so we don't modify the expression
+          case _ => IfThenElse(cond, exp1, exp2, tpe, loc)
+        }
+      }
+
+      f.copy(exp = RemoveKnownConditional.foldExpression(f.exp))
+    }
+
+  }
+
+  object CopyPropagation {
+
+    def optimize(f: Constant): Constant = {
+
+      /**
+        * Replaces any references to `oldVar` with `newVar`.
+        */
+      class ReplaceVarInExp(oldVar: Name.Ident,
+                            newVar: Name.Ident,
+                            newOffset: scala.Int,
+                            newLoc: SourceLocation) extends ExpressionFolder {
+        override def foldVar(ident: Name.Ident,
+                             offset: scala.Int,
+                             tpe: Type,
+                             loc: SourceLocation): Expression = {
+          if (ident.name == oldVar.name) {
+            Var(newVar, newOffset, tpe, newLoc)
+          } else {
+            Var(ident, offset, tpe, loc)
+          }
+        }
+
+        override def foldMkClosure(lambda: Lambda,
+                                   freeVars: List[FreeVar],
+                                   tpe: Type.Lambda,
+                                   loc: SourceLocation): Expression = {
+          val fFreeVars = freeVars.map {
+            case f @ FreeVar(ident, _, t) =>
+              if (ident.name == oldVar.name) {
+                FreeVar(newVar, newOffset, t)
+              } else {
+                f
+              }
+          }
+          MkClosure(lambda, fFreeVars, tpe, loc)
+        }
+
+        override def foldMkClosureRef(ref: Ref,
+                                      freeVars: List[FreeVar],
+                                      tpe: Type.Lambda,
+                                      loc: SourceLocation): Expression = {
+          val fFreeVars = freeVars.map {
+            case f @ FreeVar(ident, _, t) =>
+              if (ident.name == oldVar.name) {
+                FreeVar(newVar, newOffset, t)
+              } else {
+                f
+              }
+          }
+          MkClosureRef(ref, fFreeVars, tpe, loc)
+        }
+      }
+
+      object PropagateSimpleBindings extends ExpressionFolder {
+        override def foldLet(ident: Name.Ident,
+                             offset: scala.Int,
+                             rhs: Expression,
+                             body: Expression,
+                             tpe: Type,
+                             loc: SourceLocation): Expression = rhs match {
+          // If we have something like
+          //    Let x = y in body
+          // We can transform that to just
+          //    body
+          // where we've replaced all references to `x` with `y` in `body`.
+          case Var(bIdent, bOffset, _, bLoc) => new ReplaceVarInExp(ident, bIdent, bOffset, bLoc).foldExpression(body)
```

Comment:
To elaborate (and confirm that this is what @magnus-madsen is referring to), consider:

```
  let x1 = y1 in
    let x2 = y2 in
      x1 + x2
```

The innermost body `x1 + x2` is processed twice, once for each let-binding.